Wait for config entry platform forwards

bdraco opened the pull request:
Blog post https://github.com/home-assistant/developers.home-assistant/pull/1390

## Breaking change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Wait for config entry platform forwards to ensure entries are never reloaded before they are finished setting up platforms and entities.

The following integrations have not been updated as they require more complex refactoring and/or test updates and may still suffer from reload races after this PR (each one will probably need its own PR).  They need to be refactored to replace calls to `hass.config_entries.async_setup_platforms` with `hass.config_entries.async_forward_entry_setups` and/or await all `hass.config_entries.async_forward_entry_setup`

- [ ] ambient_station
- [x] axis https://github.com/home-assistant/core/pull/75178
- [x] cast - https://github.com/home-assistant/core/pull/86558
- [x] deconz - https://github.com/home-assistant/core/pull/74823
- [x] escea https://github.com/home-assistant/core/pull/86559
- [ ] esphome (race still present since it loads late)
- [ ] fjaraskupan
- [x] heos https://github.com/home-assistant/core/pull/86562
- [x] homekit_controller https://github.com/home-assistant/core/pull/75021
- [x] homematicip_cloud https://github.com/home-assistant/core/pull/86563
- [x] islamic_prayer_times https://github.com/home-assistant/core/pull/86564
- [ ] knx
- [x] konnected https://github.com/home-assistant/core/pull/86565
- [x] landisgyr_heat_meter https://github.com/home-assistant/core/pull/86569
- [x] lg_soundbar https://github.com/home-assistant/core/pull/86570
- [x] melnor https://github.com/home-assistant/core/pull/86572
- [ ] mikrotik
- [ ] nest (legacy)
- [x] openthrem_gw (https://github.com/home-assistant/core/pull/73806/#discussion_r907774078)
- [x] plex (https://github.com/home-assistant/core/pull/73806/#discussion_r907776540)
- [ ] point
- [x] samsungtv https://github.com/home-assistant/core/pull/86575
- [ ] shelly
- [ ] steam_online
- [ ] tellduslive
- [x] tomorrowio https://github.com/home-assistant/core/pull/86579
- [x] totalconnect https://github.com/home-assistant/core/pull/86580
- [x] unifi https://github.com/home-assistant/core/pull/74835
- [ ] wemo
- [ ] vesync
- [x] zha https://github.com/home-assistant/core/pull/74834
- [ ] zwave_js
- [ ] zwave_me

Alternate solution to https://github.com/home-assistant/core/pull/73613#issuecomment-1162337319 when combined with https://github.com/home-assistant/core/pull/73828


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #73654 fixes #73400
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/developers.home-assistant/pull/1390

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Wait for config entry platform forwards

bdraco committed:
tweak

bdraco committed:
await

bdraco committed:
fix device_tracker deadlock

bdraco committed:
notify as well

bdraco committed:
revert shelly

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
cleanup a few more

bdraco committed:
few more

bdraco committed:
bump isort

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
revert deconz change

bdraco committed:
revert cast

bdraco committed:
revert heos

bdraco committed:
revert homematicip_cloud

bdraco committed:
revert samsungtv

bdraco committed:
revert unifi

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
last one ..hopefully

bdraco committed:
update docstring

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
3 more

bdraco committed:
two more

bachya commented:
If I’m reading this change correctly, `ambient_station` should be easy to adjust – platform setup happens here: https://github.com/home-assistant/core/blob/9a0b3796d3fcf45b384604f10cb4367f27b0facb/homeassistant/components/ambient_station/__init__.py#L182

emontnemery commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -367,12 +367,7 @@ def async_discovered_player(
 
     async def setup_platforms_and_discovery(self):
```
This function needs to be awaited to break the race

bdraco replied:
109024fc3ce917c377615660c6f6ad61cb5c4af1

emontnemery commented on homeassistant/config_entries.py:
```diff
@@ -1133,6 +1133,14 @@ def async_setup_platforms(
         for platform in platforms:
             self.hass.async_create_task(self.async_forward_entry_setup(entry, platform))
```
Should we add a deprecation warning to `async_setup_platforms`?

bdraco replied:
a7955863f9ba6b81a7df5c1aa8a28d4dd4147169

emontnemery commented on homeassistant/components/esphome/entry_data.py:
```diff
@@ -96,13 +96,8 @@ async def _ensure_platforms_loaded(
-            if tasks:
-                await asyncio.wait(tasks)
+            if needed:
+                await hass.config_entries.async_forward_entry_setups(entry, needed)
```
This is called from an untracked task created in `async_create_task`

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco commented:
Trying to keep this fresh, but waiting until after .1 to merge it since it will make conflicts/backports more difficult if we merge before then

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
revert tomorrowio

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
drop extra function

bdraco commented on homeassistant/components/apple_tv/__init__.py:
```diff
@@ -61,12 +61,7 @@ async def on_hass_stop(event):
+        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
         await manager.init()
 
     hass.async_create_task(setup_platforms())
```
f23a511463b8de19f9862e2b8deb516450881143

bdraco committed:
drop extra function

bdraco commented on homeassistant/components/axis/device.py:
```diff
@@ -230,13 +230,8 @@ async def async_setup(self):
         self.product_type = self.api.vapix.product_type
 
         async def start_platforms():
```
bba7ed5a91e7af30718172e29aa7c5deaf3f8167

bdraco committed:
add comment about adding comp

bdraco commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -47,6 +47,9 @@ def is_on(hass: HomeAssistant, entity_id: str) -> bool:
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the device tracker."""
+
+    hass.config.components.add(DOMAIN)
```
c99798a4cee6568c82f1fedd5a7929a200123d73

bdraco committed:
remove extra function

bdraco committed:
await in sonos

bdraco committed:
remove extra function

bdraco committed:
add warning about calling async_setup_platforms

bdraco committed:
load logbook sooner for backwards compat

bdraco committed:
load logbook sooner for backwards compat

bdraco commented on homeassistant/bootstrap.py:
```diff
@@ -88,6 +88,16 @@
     # as possible so problem integrations can
     # be removed
     "frontend",
+    # Ensure the logbook integration platform helper is loaded
+    # before any logbook platforms since some legacy
+    # integration events (at least doorbird) use a different
+    # event type per device.
+    #
+    # Example: doorbird_device123_triggered
+    #
+    # Once we get rid of all
+    # the events per device, we can remove this.
+    "logbook",
```
I don't like this, but I think its OK. since don't have a better way.  

I'm still investigating an alternate solution so I may back it back out shortly

bdraco replied:
So we don't need this.  Its a strait up bug in doorbird

bdraco replied:
Fixed in https://github.com/home-assistant/core/pull/74649

bdraco committed:
revert

bdraco committed:
revert axis changes

bdraco committed:
debug

bdraco committed:
fix wind gust in tomorrowio/climacell

Fixes
```
022-07-07 12:36:20.811 ERROR    MainThread homeassistant:runner.py:96 Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/homeassistant/helpers/entity.py", line 892, in _async_registry_updated
    self.async_write_ha_state()
  File "/Users/bdraco/home-assistant/homeassistant/helpers/entity.py", line 532, in async_write_ha_state
    self._async_write_ha_state()
  File "/Users/bdraco/home-assistant/homeassistant/helpers/entity.py", line 573, in _async_write_ha_state
    attr.update(self.extra_state_attributes or {})
  File "/Users/bdraco/home-assistant/homeassistant/components/climacell/weather.py", line 153, in extra_state_attributes
    speed_convert(self.wind_gust, SPEED_MILES_PER_HOUR, self._wind_speed_unit),
  File "/Users/bdraco/home-assistant/homeassistant/util/speed.py", line 58, in convert
    raise TypeError(f"{value} is not of numeric type")
TypeError: None is not of numeric type
```

bdraco commented:
> I think [07d1ba2](https://github.com/home-assistant/core/commit/07d1ba2babc2ce13f25a0f9ee4414cc01bda5e2c) should have it's own PR if it's a bugfix?

Yes it should.  I'll break it out once I have the CI passing

bdraco committed:
fix ci

bdraco commented:
> I think [07d1ba2](https://github.com/home-assistant/core/commit/07d1ba2babc2ce13f25a0f9ee4414cc01bda5e2c) should have it's own PR if it's a bugfix?

https://github.com/home-assistant/core/pull/74660

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

bdraco committed:
Merge branch 'dev' into config_entries_wait

frenck commented:
Looks good to me! Very nice 👍
